File: project_euler/Problem004/Problem004.c
#include <stdio.h>

static int is_palindrome(int value);

void Problem004(void) {
    int target = 100;
    int result = 0;
    for (int i = 999; i >= target; --i) {
        for (int j = i; j >= target; --j) {
            int value = i * j;
            if (is_palindrome(value) && result < value) {
                result = value;
                target = j;
                break;
            }
        }
    }
    printf("%d\n", result);
}

static int is_palindrome(int value) {
    int divisor = 1;
    while (10 <= value / divisor) {
        divisor *= 10;
    }

    while (value > 1) {
        int head = value / divisor;
        int tail = value % 10;
        if (head != tail) {
            return 0;
        }

        value = value % divisor / 10;
        divisor /= 100;
    }
    return 1;
}

File: project_euler/Problem004/Problem004.js
function Problem004() {
    let target = 100
    let result = 0
    for (let i = 999; i >= target; --i) {
        for (let j = i; j >= target; --j) {
            let value = i * j
            if (isPalindrome(value) && result < value) {
                result = value
                target = j
                break
            }
        }
    }
    console.log(result)
}

function isPalindrome(value) {
    let divisor = 1
    while (10 <= value / divisor) {
        divisor *= 10
    }

    while (value > 1) {
        let head = Math.floor(value / divisor)
        let tail = value % 10
        if (head !== tail) {
            return false
        }

        value = Math.floor(value % divisor / 10)
        divisor = Math.floor(divisor / 100)
    }
    return true
}

module.exports = {
    Problem004 : Problem004
};
